rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloadRequestChain.java

```diff
@@ -149,14 +154,41 @@ public Long call() throws IOException
     }
   }
 
-  private int copyIntoCache(FSDataInputStream inputStream, FileChannel fileChannel, int length,
-                            long cacheReadStart) throws IOException
+  private long copyIntoCache(FSDataInputStream inputStream,
+          FileChannel fileChannel,
+          long cacheReadStart,
+          long cacheReadEnd,
+          byte[] remoteReadBuffer) throws IOException
   {
     long start = System.nanoTime();
-    byte[] buffer = new byte[length];
+
+    long length = cacheReadEnd - cacheReadStart;
     log.debug(String.format("Copying data of file %s of length %d from position %d", remotePath, length, cacheReadStart));
-    inputStream.readFully(cacheReadStart, buffer, 0, length);
+    if (length <= remoteReadBuffer.length) {
+      inputStream.readFully(cacheReadStart, remoteReadBuffer, 0, Math.toIntExact(length));
+      writeToFile(remoteReadBuffer, Math.toIntExact(length), fileChannel, cacheReadStart);
+    }
+    else {
+      // Use streaming reads here as we will be doing multiple iterations
+      long leftToRead = length;
+      while (leftToRead > 0) {
+        int toRead = Math.toIntExact(Math.min(remoteReadBuffer.length, leftToRead));
```

Comment:
can't we use readFully here as well with updated caceReadStart in every iteration?

Reply:
also, should we parallelize this?

Reply:
streaming read() provides read-ahead benefit that can be leveraged in this case as we know that we will be doing sequential read here.

Not need to parallelize this further, we need to keep the impact of parallel warmup minimal.